testng-core/src/test/java/test/annotationtransformer/AnnotationTransformerTest.java

```diff
@@ -270,4 +276,51 @@ public void ensureAnnotationTransformerIsNotInvokedMultipleTimes() {
     int retried = RetryListener.getExecutionCount();
     assertThat(retried).isEqualTo(1);
   }
+
+  @Test(description = "GITHUB-2536", dataProvider = "getXmlPackages")
+  public void testNestedTestClass_packageIncludesNormalTestClasses(List<XmlPackage> xmlPackages) {
+    ExternalGroupManager groupManager = new ExternalGroupManager();
+    String group = "run-me-dynamically";
+    String container = ClassContainer.class.getName();
+    String inner = NonGroupClass3.class.getSimpleName();
+
+    groupManager.addGroupDefinition(container + "$" + inner, group);
+
+    TestListenerAdapter tla = new TestListenerAdapter();
+
+    XmlSuite xmlSuite = createXmlSuite("2536_suite");
+    XmlTest myTest = createXmlTest(xmlSuite, "2536_test");
+    myTest.addIncludedGroup(group);
+    TestNG testng = create(xmlSuite);
+    testng.addListener(tla);
+    testng.addListener(groupManager);
+
+    myTest.setXmlPackages(xmlPackages);
+
+    testng.run();
+    assertThat(tla.getFailedTests()).withFailMessage("We should have no failed tests").isEmpty();
+
+    assertThat(tla.getPassedTests()).withFailMessage("We should have 2 passed tests").hasSize(2);
+
+    long count =
+        tla.getPassedTests().stream()
+            .filter(m -> m.getInstance().getClass().equals(NonGroupClass3.class))
+            .count();
+
+    assertThat(count)
+        .withFailMessage("We should have tests that only belong to " + NonGroupClass3.class)
+        .isEqualTo(2);
+  }
+
+  @DataProvider
+  public Object[][] getXmlPackages() {
```

Comment:
What is the difference between the 2 cases?

Reply:
@juherr none to be honest. I just added the failing tests from the defect for completeness.